Fix cumulative issues

utjkaz opened the pull request:
- Fix class name conflicts between dll and player (e.g. class Mutex) 
I've added namespace AudioPluginUtil. Because "class Mutex" has cuased name conflicts between dll & player. In some platform, it'll cause some crashes.

- Fix new / delete mismatching(effectdata is correct in some case)
I've found new / delete mismatching.  (data <> effectdata)

- Fix compiler warning about numeric casting (int to float, int to unsigned int, ...)
I'd prefer to fix it because it'll dismiss important notices.

- Support PlatformMutex
I've added to support PlatformMutex for any other platforms. This future is disabled by default.

utjkaz committed:
- Fix class name conflicts between dll and player (e.g. class Mutex)
- Fix new / delete mismatching(effectdata is correct in some case)
- Fix compiler warning about numeric casting
- Support PlatformMutex

utjkaz committed:
- Fix compiler warnings about casting
- Revert including assert.h just in case

Tak commented on NativeCode/AudioPluginUtil.cpp:
```diff
@@ -26,7 +30,7 @@ template<typename T>
 {
     unsigned int count = 1, numbits = 0;
-    while (count < numsamples)
+    while (count < (unsigned)numsamples)
```
I would make the locals ints here instead of potentially forcing `numsamples` to overflow

utjkaz replied:
Okay, I'll resolve it with casting to int.

```suggestion
    while ((int)count < numsamples)
```

utjkaz replied:
Okay, I've fixed this casting.

Tak commented on NativeCode/AudioPluginUtil.cpp:
```diff
@@ -257,6 +261,7 @@ void HistoryBuffer::ReadBuffer(float* buffer, int numsamplesTarget, int numsampl
     buffer[numsamplesTarget] = (float)n; // how many samples were written
 }
 
+#ifndef PLATFORM_MUTEX
```
Maybe we should extract `Mutex` to its own source files so we don't have so many ifdef blocks

utjkaz replied:
Okay, I'll split the mutex class. (It's better to customize it, especially platform dependent.)

utjkaz replied:
Okay, I've revertes changes for PLATFORM_MUTEX.

Tak commented on NativeCode/AudioPluginUtil.h:
```diff
@@ -493,3 +505,7 @@ void DeclareEffect(
+
+} // namespace AudioPluginUtil
+
+using namespace AudioPluginUtil;
```
:grimacing: Is this to avoid making users change their code?
If it's broken today, I'd prefer to have users update their plugins than to have this using statement in the header.

utjkaz replied:
This fix is necessary to resolve class name conflicts between player code and dlls. And partially yes, it will avoid user define classes on global namespace. Actually, "class Mutex" seems conflicts bettwen player code and dlls. In some case, it'll cause the issue about function call mismatching.
(e.g.)
When the constructor of Mutex(Mutex::Mutex) will be called into dlls, Mutex::Mutex in player code will be called. (However, it depends on the platform.)

The one of the other solutions is renaming class name Thread to HogeThread. However we should check all class names on global namespace.

Tak replied:
Right, I think I wasn't clear, sorry.
I understand why `Mutex` got moved into the new namespace.
The part I don't like is the `using` statement in the header - I think the code that uses this should import the namespace or explicitly use `AudioPluginUtil::Mutex`

utjkaz replied:
Okay, I'll remove `using namespace AudioPluginUtil` in the header & it'll be moved on .each cpp file.

utjkaz replied:
Okay, I've added the "AudioPluginUtil::" prefix for all referenced classes/functions/constants.

utjkaz committed:
Add AudioPluginUtil namespace for all referenced classes, functions and constants.

Add fixed value defines for Android & any other platforms in AudioPluginInterface.h.

Revert changes for PLATFORM_MUTEX

Tak commented on NativeCode/Plugin_ConvolutionReverb.cpp:
```diff
@@ -50,15 +50,15 @@ namespace ConvolutionReverb
-        UnityComplexNumber** h;
-        UnityComplexNumber** x;
+        AudioPluginUtil::UnityComplexNumber** h;
+        AudioPluginUtil::UnityComplexNumber** x;
```
You could also have added `using` statements in the cpp files, but I think it's fine either way
